### DISCUSSION TITLE
How to Implement Backend Conditions in User Sign In

### DISCUSSION THREAD

**Discussion by matanshamshila** [pocketbase/pocketbase#4073] [Q&A] [answered]
Hello everyone,

I work in a company with a closed network, and I am using PB in a web app. I have a URL that allows me to fetch the name of the currently connected user on the PC. My goal is that the user will be able to sign in only if the inputted username matches the one fetched from the URL.

To achieve this, I cannot solely depend on the frontend. I need a more robust solution that validates this on the backend.

I would appreciate any suggestions or guidance on how I might do it

**Comment by jonshipman** [comment]
For authentication, you need a secret on the server and client and a way to verify.

The URL where you can fetch the currently connected user - how does that work? If that's the source of truth, can't it also produce a JWT that can be passed to your application? Then the client would just need to check the URL and database and compare them.

Alternative: just identify people by their email. Setup a magic link and have the click in the email log that person in.

**Comment by matanshamshila** [comment]
I am not sure how to authenticate users in pocketbase based on token and not the standard authWithPassword function.
I want more smooth and simple solution and I prefer to not use email for sign in

I will explain it more clearly

1. Each person in the company logs in to a windows computer with a username and a password.
I have url like example.com/auth, that I can fetch from the username of the user currently connected to that PC.
I receive a JSON object, one of its properties is the current username.
the username is like the employee id, It's the same data as typing "whoami" in the cmd.

2. I have the standard sign in function in pocketbase (attached below).
Currently I save the username I fetched in a variable, and I sign in with this variable.
the problem people can manipulate the frontend, ignore the fetch function and put a different username in the variable.
I am trying to come up with a solution that will enforces users to connect only with the username I fetched,
and that they will not be able to simply edit the frontend and connect with a different username.

const authData = await pb.collection('users').authWithPassword(
    'YOUR_EMAIL',
    'YOUR_PASSWORD',
);

What do you think?

**Reply by ganigeorgiev** [reply, marked as answer]
I don't understand what is the desired workflow, but here is a sample custom auth route using the JS `pb_hooks` that will send an http request to fetch the user data from a trusted source and then return a PocketBase auth response.

1. Register a new POST route for example with the following `pb_hooks/main.pb.js` file:
    ```js
    /// <reference path="../pb_data/types.d.ts" />

    // For more details check:
    // https://pocketbase.io/docs/js-routing/
    // https://pocketbase.io/docs/js-sending-http-requests/
    // https://pocketbase.io/docs/js-records/
    routerAdd("POST", "/auth/:username", (c) => {
        let username = c.pathParam("username")

        const result = $http.send({
            url:     "https://example.com/your/trusted/auth/url",
            method:  "POST",
            body:    JSON.stringify({ "username": username }), // eg. pass the username or other data as json body if needed
            headers: { "content-type": "application/json" },
        })

        if (result.statusCode != 200 || result.json.username != username) {
            throw new ForbiddenError("You are not allowed to authorize with " + username);
        }

        // find a PocketBase user with the authorized username
        const user = $app.dao().findAuthRecordByUsername("users", username)

        return $apis.recordAuthResponse($app, c, user);
    })
    ```
2. Then on the client-side to send request to this new route you can use the `send()` SDK helper method:
   ```js
   const result = await pb.send("/auth/someusername", {
        method: "POST",
    });

    // store the token and user in the auth store
    pb.authStore.save(result.token, result.record)

    // any following requests will have Authorization: TOKEN from the previous stored state...
    const items = await pb.collection("example").getFullList();
   ```

**Reply by matanshamshila** [reply]
Thank you very much for the help
This approach works except the part of $http.send

I have tried to fetch the data from the real URL I am using many times, and I get an empty object
It's probably not due to CORS policy, because simple fetch operation from the frontend works fine
I have also tried to fetch data from other URLs (like https://openlibrary.org/books/OL11994542M.json), and It's not working.
How I might solve it?

Here is my code after I isolated the part of $http.send
(the rest of the code works except the fetch that returns an empty object)

```javascript
routerAdd("POST", "/auth/:username", (c) => {
    let username = c.pathParam("username")

    const result = $http.send({
        url:     "https://openlibrary.org/books/OL11994542M.json",
        method:  "GET",
        headers: { "content-type": "application/json" },
    })

    // if (res.statusCode != 200 || result.json.username != username) {
    //     throw new ForbiddenError("You are not allowed to authorize with " + username);
    // }

    // find a PocketBase user with the authorized username
    // const user = $app.dao().findAuthRecordByUsername("users", username)

    // return $apis.recordAuthResponse($app, c, user);

    return result.json;
})
```

```javascript
 const result = await pb.send("/auth/someusername", {
        method: "POST",
    });

    console.log(result);
```

**Reply by ganigeorgiev** [reply]
Returning `result.json` is incorrect. If you want to return the json to the user you need to write it to the response (see https://pocketbase.io/docs/js-routing/#writing-response-body):

```js
return c.json(200, result.json)
```

**Reply by matanshamshila** [reply]
This code works as described
this is the solution I was looking for
thank you a lot, really appreciate it!

just there might be a small typo here. seems like it should be "result" instead of "res'

```javascript
if (result.statusCode != 200 || result.json.username != username) {
```
